File: backend/src/controllers/registrationController.ts
import asyncHandler from 'express-async-handler';
import type { Request, Response } from 'express';
import { stringify } from 'csv-stringify/sync';
import Event from '../models/Event.js';
import Registration from '../models/Registration.js';
import { ok, fail } from '../utils/apiResponse.js';
import { sendEmail } from '../utils/sendEmail.js';

export const registerForEvent = asyncHandler(async (req: Request, res: Response) => {
  if (!req.user) return fail(res, 'Not authorized', 401);

  const event = await Event.findById(req.params.eventId);
  if (!event || !event.isActive) return fail(res, 'Event not found', 404);

  if (event.isCancelled) return fail(res, 'Event is cancelled', 400);

  if (event.registrationDeadline && new Date() > event.registrationDeadline) {
    return fail(res, 'Registration deadline has passed', 400);
  }

  if (event.registeredCount >= event.capacity) {
    return fail(res, 'Event is full', 400);
  }

  const existing = await Registration.findOne({ student: req.user._id, event: event._id });
  if (existing && existing.status !== 'cancelled') {
    return fail(res, 'Already registered', 409);
  }

  let registration;
  if (existing && existing.status === 'cancelled') {
    existing.status = 'registered';
    existing.registeredAt = new Date();
    existing.checkedInAt = undefined;
    registration = await existing.save();
  } else {
    registration = await Registration.create({
      student: req.user._id,
      event: event._id,
      status: 'registered'
    });
  }

  event.registeredCount += 1;
  await event.save();

  await sendEmail({
    to: req.user.email,
    subject: `Registration Confirmed: ${event.title}`,
    text: `You have successfully registered for ${event.title} on ${event.date.toISOString()}.`
  });

  return ok(res, registration, 'Registered successfully', 201);
});

export const cancelOwnRegistration = asyncHandler(async (req: Request, res: Response) => {
  if (!req.user) return fail(res, 'Not authorized', 401);

  const registration = await Registration.findOne({
    student: req.user._id,
    event: req.params.eventId,
    status: { $in: ['registered', 'attended'] }
  });

  if (!registration) return fail(res, 'Registration not found', 404);

  registration.status = 'cancelled';
  await registration.save();

  await Event.findByIdAndUpdate(registration.event, { $inc: { registeredCount: -1 } });
  return ok(res, null, 'Registration cancelled');
});

export const getMyRegistrations = asyncHandler(async (req: Request, res: Response) => {
  if (!req.user) return fail(res, 'Not authorized', 401);

  const rows = await Registration.find({ student: req.user._id, status: { $ne: 'cancelled' } })
    .populate('event')
    .sort({ registeredAt: -1 });

  return ok(res, rows);
});

export const getEventRegistrants = asyncHandler(async (req: Request, res: Response) => {
  const rows = await Registration.find({ event: req.params.eventId, status: { $ne: 'cancelled' } })
    .populate('student', 'name email rollNumber department')
    .sort({ registeredAt: -1 });

  return ok(res, rows);
});

export const markAttendance = asyncHandler(async (req: Request, res: Response) => {
  const registration = await Registration.findById(req.params.id);
  if (!registration) return fail(res, 'Registration not found', 404);

  registration.status = 'attended';
  registration.checkedInAt = new Date();
  await registration.save();

  return ok(res, registration, 'Attendance marked');
});

export const exportParticipantsCsv = asyncHandler(async (req: Request, res: Response) => {
  const rows = await Registration.find({ event: req.params.eventId, status: { $ne: 'cancelled' } })
    .populate('student', 'name email rollNumber department')
    .populate('event', 'title date venue');

  const data = rows.map((row: any) => ({
    name: row.student?.name || '-',
    email: row.student?.email || '-',
    rollNumber: row.student?.rollNumber || '-',
    department: row.student?.department || '-',
    status: row.status,
    registeredAt: row.registeredAt,
    checkedInAt: row.checkedInAt || ''
  }));

  const csv = stringify(data, { header: true });
  res.setHeader('Content-Type', 'text/csv');
  res.setHeader('Content-Disposition', 'attachment; filename="participants.csv"');
  res.send(csv);
});

File: backend-spring/src/main/java/com/campusconnect/controller/RegistrationController.java
package com.campusconnect.controller;

import com.campusconnect.dto.ApiResponse;
import com.campusconnect.model.Registration;
import com.campusconnect.model.User;
import com.campusconnect.service.RegistrationService;
import com.opencsv.CSVWriter;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.io.IOException;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/registrations")
@RequiredArgsConstructor
public class RegistrationController {

    private final RegistrationService registrationService;

    @PostMapping("/{eventId}")
    public ResponseEntity<ApiResponse<Registration>> register(
            @AuthenticationPrincipal User user,
            @PathVariable String eventId) {
        Registration reg = registrationService.registerForEvent(user, eventId);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.ok(reg, "Registered successfully"));
    }

    @DeleteMapping("/{eventId}")
    public ResponseEntity<ApiResponse<Void>> cancel(
            @AuthenticationPrincipal User user,
            @PathVariable String eventId) {
        registrationService.cancelOwnRegistration(user, eventId);
        return ResponseEntity.ok(ApiResponse.ok(null, "Registration cancelled"));
    }

    @GetMapping("/my")
    public ResponseEntity<ApiResponse<List<Map<String, Object>>>> myRegistrations(
            @AuthenticationPrincipal User user) {
        return ResponseEntity.ok(ApiResponse.ok(registrationService.getMyRegistrations(user)));
    }

    @GetMapping("/event/{eventId}")
    public ResponseEntity<ApiResponse<List<Map<String, Object>>>> eventRegistrants(
            @PathVariable String eventId) {
        return ResponseEntity.ok(ApiResponse.ok(registrationService.getEventRegistrants(eventId)));
    }

    @PutMapping("/{id}/attend")
    public ResponseEntity<ApiResponse<Registration>> markAttendance(@PathVariable String id) {
        Registration reg = registrationService.markAttendance(id);
        return ResponseEntity.ok(ApiResponse.ok(reg, "Attendance marked"));
    }

    @GetMapping("/event/{eventId}/export")
    public void exportCsv(@PathVariable String eventId, HttpServletResponse response) throws IOException {
        response.setContentType("text/csv");
        response.setHeader("Content-Disposition", "attachment; filename=\"participants.csv\"");

        List<Map<String, Object>> data = registrationService.getExportData(eventId);

        try (CSVWriter writer = new CSVWriter(response.getWriter())) {
            // Header
            writer.writeNext(new String[]{"name", "email", "rollNumber", "department", "status", "registeredAt", "checkedInAt"});

            for (Map<String, Object> row : data) {
                writer.writeNext(new String[]{
                        str(row.get("name")),
                        str(row.get("email")),
                        str(row.get("rollNumber")),
                        str(row.get("department")),
                        str(row.get("status")),
                        str(row.get("registeredAt")),
                        str(row.get("checkedInAt"))
                });
            }
        }
    }

    private String str(Object obj) {
        return obj != null ? obj.toString() : "";
    }
}
